### ISSUE TITLE
Use the Lobby as a portal from one hubs room to another

### ISSUE THREAD

**Issue by blairmacintyre** [mozilla/hubs#1999]
**Is your feature request related to a problem? Please describe.**
The problem of how to support large audiences for a talk / event is critical to supporting things like distributed events and conferences.  How can we have many people watching what's happening in a hubs room, even if it's limited to 25 people or so?  

This also applies to our conference scenario with streaming video;  even though the video can be replicated across many rooms to deal with load, if we want to have in-hubs moderators or paper co-authors, they need to be in one room, and it would be nice if all attendees had "access".

One option discussed is having a small number in the room, but having lots remain in something like the lobby.  They can send chat messages, see what's happening from a predefined view, etc.  But that loses the social aspect. 

**Describe the solution you'd like**
Building on that, what if we could put the lobby view on the wall (or, rather, "through the wall") of a room.  From the viewpoint of the central room, it's no different (people "aren't there" and would be included in the lobby count, and could send messages in via chat).

The new rooms would see a view through a "portal" or "hole in the way" that corresponds to what they would have seen in the lobby, but are in a room with other people, and can move around and talk and so forth.  The view through the portal/hole would respect 3D, since it's being rendered locally (as the lobby would be):  the lobby viewpoint would be attached to the wall / portal, so audio coming it would be adjusted as appropriate, and so on.

The negative is that it increases the load on a local machine (the machine now has to render the content and avatars from both rooms).  But, if we adjusted the limits (e.g., N people in the joint room and (25-N) in the rooms looking in) the total load would be similar.

It's like you're in a theatre, and everyone is in box seating with 10-15 people per box.  

**Describe alternatives you've considered**
The hubs team has considered a bunch of alternatives to the "large scene" situation.

**Comment by gfodor** [comment]
This sounds like a specific UX on top of the general model of room
sharding. This can be accomplished either via:

- room mirroring, where the server transmits data from more than one room
to clients

- client multi-room, where the client connects to more than one room
simultaneously

In general I think we need to avoid adding the ability for a client to be
consuming data from multiple rooms through either mechcanism, unless we are
absolutely forced to. The reason is that this general facility has a huge
blast radius in terms of complexity.

For example, this introduces the need for a new abstraction above the room
level, to bind the various channels. It also means the client code has to
be able to deal with things at a higher level of abstraction, where
messages can’t be assumed to come from the same room. Also, the whole
security system currently is room-scoped, so that would need to be rebuilt
or hacked around. Also, there is the eventual problem of routing: if this
system is intended to help spread load, you would need to implement the
requisite load balancing system to ensure users don’t all land on the same
server.

I think the thing we need to do first is to try to optimize the room-local
lobby experience so it stays intra-room but provides a good experience for
a few hundred CCU. The main constraint is that clients will not be able to
process or render more than two dozen articulated avatars/voices, so that’s
a pretty rich surface area to explore.

**Comment by gfodor** [comment]
Also, the current lobby experience is currently limited only in part due to
the performance constraints. We also deliberately have limited camera
control and other agency to lobby participants because of the social system
trade off that would result where room participants do not have visibility
into the viewpoint of those in the lobby. For example, we didn’t want full
controls in the lobby since room participants may feel they are being spied
upon or are fearful that lobby participants are viewing them or listening
to them from undesired vantage points.

For the conference case, I think in combinations with seats, we can create
a new mode for the “lobby” where people logically join the room, are seen
as avatars in seats but have no locomotion or body language, and are muted
unless called upon.  In that case the net load on the client is nearly the
same as the lobby if we can batch the avatars, and by putting people in
seats we can give them a better sense of copresence.

**Comment by gfodor** [comment]
any other ways we can improve the UX for lobby? also, we are planning on making it possible to show a ranked list of rooms on the homepage, so that was intended to help solve the case of capacity sharding.

**Comment by gfodor** [comment]
Are there any other ways we can improve the lobby experience for
conferences to help increase the value of it for audience members? We could
also perhaps allow a setting for room owners to allow lobby locomotion. We
could, in the extreme case, allow everyone to enter the room from a UX
perspective but anyone joining above the cap would be invisible to others
(this obviously sucks from a communication perspective, but is better than
being locked in place)

**Comment by gfodor** [comment]
Are there any other ways we can improve the lobby experience for
conferences to help increase the value of it for audience members? We could
also perhaps allow a setting for room owners to allow lobby locomotion. We
could, in the extreme case, allow everyone to enter the room from a UX
perspective but anyone joining above the cap would be invisible to others
(this obviously sucks from a communication perspective, but is better than
being locked in place)

**Comment by blairmacintyre** [comment]
From my experience so far, I think some variation of the stadium seating would be sufficient IF people could choose which "group" they are in in the stadium and be able to chat with those nearby.  If we have a situation where people are just watching a room, but not interacting, we may as well just broadcast the room as a video on twitch, gateway the twitch chat to/from the room, and skip all the lobby complexity.

**Comment by gfodor** [comment]
yeah, that would basically be another variant of room sharding, since it means that those users would need to be consuming multiple sets of webrtc peers, etc. i think the path for that right now is going to be creating N rooms and putting them on the homepage. if people want to join a specific group, they can either choose a room by name to coordinate on, or create their own private room and share it async. and in this case you will only be able to co-watch video, not a 3d mirror of an existing room.

**Comment by gfodor** [comment]
also, i would argue that if you are actually going to be streaming a 3d room in hubs, the lobby experience with chat + locomotion would be superior to twitch streaming video, because the users would have camera controls and potentially other affordances we could add over time to allow engagement with the speaker beyond text chat.

**Comment by gfodor** [comment]
one could also imagine eventually building a simple Q&A mechanic into this, where the lobby participants can be 'called on' to join the room (and bypass the cap limit) for a short duration.

the main thing you're losing to this relative to a more complex system doing room sharding is the ability for audience members to communicate with one another as avatars. in practice, mirroring has its own distinct tradeoffs -- often audience members will be confused about the fact they can not be heard/seen by all other avatars in the mirrored case.

**Comment by gfodor** [comment]
(specifically, if you only mirror the person/people presenting, audience members will have to 'learn' that those avatars, despite appearing very real, will not see or hear them at all, and will basically appear to be dumb recordings.)
